homeassistant/components/zha/radio_manager.py

```diff
@@ -156,3 +212,171 @@ async def async_reset_adapter(self) -> None:
+        ):
+            # No point in prompting the user if the backup doesn't have a new IEEE
+            # address or if there is no way to overwrite the IEEE address a second time
+            await self.restore_backup(self.chosen_backup)
+
+            return False
+
+        return True
+
+    async def async_restore_backup_step_2(self, overwrite_ieee: bool) -> None:
+        """Restore backup and optionally overwrite IEEE."""
+        assert self.chosen_backup is not None
+
+        backup = self.chosen_backup
+
+        if overwrite_ieee:
+            backup = _allow_overwrite_ezsp_ieee(backup)
+
+        # If the user declined to overwrite the IEEE *and* we wrote the backup to
+        # their empty radio above, restoring it again would be redundant.
+        await self.restore_backup(backup)
+
+
+class ZhaMultiPANMigrationHelper:
+    """Helper class for automatic migration when upgrading the firmware of a radio.
+
+    This class is currently only intended to be used when changing the firmware on the
+    radio used in the Home Assistant Sky Connect USB stick and the Home Asssistant Yellow
+    from Zigbee only firmware to firmware supporting both Zigbee and Thread.
+    """
+
+    def __init__(
+        self, hass: HomeAssistant, config_entry: config_entries.ConfigEntry
+    ) -> None:
+        """Initialize MigrationHelper instance."""
+        self._config_entry = config_entry
+        self._hass = hass
+        self._radio_mgr = ZhaRadioManager()
+        self._radio_mgr.hass = hass
+
+    async def async_initiate_migration(self, data: dict[str, Any]) -> bool:
+        """Initiate ZHA migration.
+
+        The passed data should contain:
+        - Discovery data identifying the device being firmware updated
+        - Discovery data for connecting to the device after the firmware update is
+          completed.
+
+        Returns True if async_finish_migration should be called after the firmware
+        update is completed.
+        """
+        migration_data = HARDWARE_MIGRATION_SCHEMA(data)
+
+        name = migration_data["new_discovery_info"]["name"]
+        new_radio_type = ZhaRadioManager.parse_radio_type(
+            migration_data["new_discovery_info"]["radio_type"]
+        )
+        old_radio_type = ZhaRadioManager.parse_radio_type(
+            migration_data["old_discovery_info"]["radio_type"]
+        )
+
+        new_device_settings = new_radio_type.controller.SCHEMA_DEVICE(
+            migration_data["new_discovery_info"]["port"]
+        )
+        old_device_settings = old_radio_type.controller.SCHEMA_DEVICE(
+            migration_data["old_discovery_info"]["port"]
+        )
+
+        if (
+            self._config_entry.data[CONF_DEVICE][CONF_DEVICE_PATH]
+            != old_device_settings[CONF_DEVICE_PATH]
+        ):
+            # ZHA is using another radio, do nothing
+            return False
+
+        try:
+            await self._hass.config_entries.async_unload(self._config_entry.entry_id)
```

Comment:
I'd update the config entry data and reload the config entry instead. The updated data should trigger the config entry to retry setup by raising ConfigEntryNotReady. When we're done we can update the config entry data again, to remove the flag.

Just unloading here is fragile, because we're not guarding against something else setting up the config entry.

Reply:
Do you mean intentionally giving it an invalid value to prevent the integration from loading at all?

Reply:
Store a boolean in the config entry data which check causes the integration to raise `ConfigEntryNotReady`.

Reply:
I was considering an approach like that, but it seems really fragile. For example, if Home Assistant is shut down or crashes while migration is in progress, who is responsible for clearing the bool flag stored in the config entry data?
Maybe we could just store something in `hass.data`, so it doesn't persist across core restarts?

Also, when we're migrating to multi-PAN radio, which is what this PR supports, there's no need to restrict ZHA from trying to connect to the radio before the addon's socket is opened. It will be a concern when we add support for reverting to Zigbee only software, if ZHA tries to connect to the radio too early it may interfere with the firmware flashing.

Reply:
Then I think we should try to expose a property that says if there's a migration in progress similar to the add-on manager. The integration could then check that property and raise `ConfigEntryNotReady` if there's a migration in progress.

Reply:
That sounds very good, but I think we should do it in a separate PR which changes the logic both for the option flow and for the automatic migration.